libs/langchain/langchain/chains/conversation/base.py

```diff
@@ -10,9 +11,87 @@
 from langchain.memory.buffer import ConversationBufferMemory
 
 
+@deprecated(
+    since="0.2.7",
+    alternative=(
+        "RunnableWithMessageHistory: "
+        "https://api.python.langchain.com/en/latest/runnables/langchain_core.runnables.history.RunnableWithMessageHistory.html"  # noqa: E501
+    ),
+    removal="1.0",
+)
 class ConversationChain(LLMChain):
     """Chain to have a conversation and load context from memory.
 
+    This class is deprecated in favor of ``RunnableWithMessageHistory``. Please refer
+    to this tutorial for more detail: https://python.langchain.com/v0.2/docs/tutorials/chatbot/
+
+    ``RunnableWithMessageHistory`` offers several benefits, including:
+
+    - Stream, batch, and async support;
+    - More flexibile memory handling, including the ability to manage memory
+      outside the chain;
+    - Support for multiple threads.
+
+    Below is a minimal implementation, analogous to using ``ConversationChain`` with
+    the default ``ConversationBufferMemory``:
+
+        .. code-block:: python
+
+            from langchain_core.chat_history import InMemoryChatMessageHistory
+            from langchain_core.runnables.history import RunnableWithMessageHistory
+            from langchain_openai import ChatOpenAI
+
+
+            store = {}  # memory is maintained outside the chain
+
+            def get_session_history(session_id: str) -> InMemoryChatMessageHistory:
+                if session_id not in store:
+                    store[session_id] = InMemoryChatMessageHistory()
+                return store[session_id]
+
+            llm = ChatOpenAI(model="gpt-3.5-turbo-0125")
+
+            chain = RunnableWithMessageHistory(llm, get_session_history)
+            chain.invoke(
+                "Hi I'm Bob.",
+                config={"configurable": {"session_id": "1"}},
+            )  # session_id determines thread
+    Memory objects can also be incorporated into the ``get_session_history`` callable:
+
+        .. code-block:: python
+
+            from langchain.memory import ConversationBufferWindowMemory
+            from langchain_core.chat_history import InMemoryChatMessageHistory
+            from langchain_core.runnables.history import RunnableWithMessageHistory
+            from langchain_openai import ChatOpenAI
+
+
+            store = {}  # memory is maintained outside the chain
+
+            def get_session_history(session_id: str) -> InMemoryChatMessageHistory:
+                if session_id not in store:
+                    store[session_id] = InMemoryChatMessageHistory()
+                    return store[session_id]
+
+                memory = ConversationBufferWindowMemory(
```

Comment:
Do we need to update the memory class after invocation for other types of memory?

Reply:
If not this is actually not terrible